std-bits/table/src/main/java/org/enso/table/data/column/operation/cast/CastOperation.java

```diff
@@ -56,4 +65,422 @@ private static StorageConverter<?> fromStorageType(StorageType<?> storageType) {
+      // If we cannot convert the value to long, we return the original type.
+      return bigIntegerType;
+    }
+  }
+
+  private static StorageType<?> inferFloatType(
+      ColumnStorage<?> columnStorage, PreciseTypeOptions options) {
+    if (!options.wholeFloatsBecomeIntegers()) {
+      return columnStorage.getType();
+    }
+
+    if (!(columnStorage.getType() instanceof FloatType floatType)) {
+      throw new IllegalArgumentException(
+          "Cannot infer float type from non-integer storage: " + columnStorage.getType());
+    }
+
+    // Build the min and max of values in the column.
+    try {
+      var accumulator = new LongAccumulator();
+      StorageIterators.buildOverDoubleStorage(
+          floatType.asTypedStorage(columnStorage),
+          false,
+          floatType.makeBuilder(0, BlackholeProblemAggregator.INSTANCE),
+          (builder, index, item, isNothing) -> {
+            if (isNothing) {
+              return;
+            }
+            if (item % 1 != 0 || !IntegerType.INT_64.fits(item)) {
+              throw new ArithmeticException(
+                  "Value is not a whole number or doesn't fit in a long: " + item);
+            }
+            accumulator.accumulate((long) item, isNothing);
+          });
+
+      return accumulator.getCount() == 0
+          ? floatType
+          : (options.shrinkIntegers() ? accumulator.resolveType() : IntegerType.INT_64);
+    } catch (ArithmeticException e) {
+      // If we cannot convert the value to long, we return the original type.
+      return floatType;
+    }
+  }
+
+  private static class BigDecimalAccumulator extends LongAccumulator {
+    private boolean overflowed = false;
+
+    public void accumulate(BigDecimal item) {
+      if (item == null) {
+        return;
+      }
+
+      var bigIntegerValue = item.toBigIntegerExact();
+      if (!overflowed) {
+        if (IntegerType.INT_64.fits(bigIntegerValue)) {
+          super.accumulate(bigIntegerValue.longValueExact(), false);
+        } else {
+          overflowed = true;
+        }
+      }
+    }
+
+    public boolean getOverflowed() {
+      return overflowed;
+    }
+  }
+
+  private static StorageType<?> inferBigDecimalType(
+      ColumnStorage<?> columnStorage, PreciseTypeOptions options) {
+    if (!options.wholeFloatsBecomeIntegers()) {
+      return columnStorage.getType();
+    }
+
+    if (!(columnStorage.getType() instanceof BigDecimalType bigDecimalType)) {
+      throw new IllegalArgumentException(
+          "Cannot infer decimal type from non-decimal storage: " + columnStorage.getType());
+    }
+
+    // Build the min and max of values in the column.
+    try {
+      var accumulator = new BigDecimalAccumulator();
+      StorageIterators.buildOverStorage(
+          bigDecimalType.asTypedStorage(columnStorage),
+          false,
+          bigDecimalType.makeBuilder(0, BlackholeProblemAggregator.INSTANCE),
+          (builder, index, item) -> accumulator.accumulate(item));
```

Comment:
Why are we using a builder to just iterate over a storage?

It is handy but it unnecessarily allocates O(N) of storage for a result that is thrown away, am I right?